pkg/plugin/windows/pktmon/pktmon_plugin_windows.go

```diff
@@ -111,103 +116,135 @@ func (p *Plugin) RunPktMonServer() error {
+	// run the getflows loop
+	for {
+		err := p.GetFlow(ctx, eventCh)
+		if _, ok := status.FromError(err); ok {
+			p.l.Error("failed to get flow, retriable:", zap.Error(err))
+			continue
+		}
+		return fmt.Errorf("failed to get flow, unrecoverable: %w", err)
+	}
+}
+
+func (p *Plugin) SetupStream() error {
+	var err error
 	fn := func() error {
 		p.l.Info("creating pktmon client")
-		client, err := NewClient()
+		p.grpcClient, err = newGRPCClient()
 		if err != nil {
 			return fmt.Errorf("failed to create pktmon client before getting flows: %w", err)
 		}
 
-		str, err = client.GetFlows(ctx, &observerv1.GetFlowsRequest{})
-		if err != nil {
-			return fmt.Errorf("failed to open pktmon stream: %w", err)
-		}
 		return nil
 	}
-	err := utils.Retry(fn, connectionRetryAttempts)
+	err = utils.Retry(fn, connectionRetryAttempts)
 	if err != nil {
 		return fmt.Errorf("failed to create pktmon client: %w", err)
 	}
 
-	for {
-		select {
-		case <-ctx.Done():
-			return fmt.Errorf("pktmon context cancelled: %w", ctx.Err())
-		default:
-			err := p.GetFlow(str)
-			if err != nil {
-				return fmt.Errorf("failed to get flow from observer: %w", err)
-			}
-		}
-	}
+	return nil
 }
 
-func (p *Plugin) GetFlow(str observerv1.Observer_GetFlowsClient) error {
-	event, err := str.Recv()
-	if err != nil {
-		return fmt.Errorf("failed to receive pktmon event: %w", err)
+func (p *Plugin) StartStream(ctx context.Context) error {
+	if p.grpcClient == nil {
+		return fmt.Errorf("unable to start stream: %w", ErrNilGrpcClient)
 	}
 
-	fl := event.GetFlow()
-	if fl == nil {
-		p.l.Error("received nil flow, flow proto mismatch from client/server?")
+	var err error
+	fn := func() error {
+		p.stream, err = p.grpcClient.GetFlows(ctx, &observerv1.GetFlowsRequest{})
+		if err != nil {
+			return fmt.Errorf("failed to open pktmon stream: %w", err)
+		}
 		return nil
 	}
-
-	ev := &v1.Event{
-		Event:     fl,
-		Timestamp: fl.GetTime(),
+	err = utils.Retry(fn, connectionRetryAttempts)
+	if err != nil {
+		return fmt.Errorf("failed to create pktmon client: %w", err)
 	}
 
-	if fl.GetType() == flow.FlowType_L7 {
-		dns := fl.GetL7().GetDns()
-		if dns != nil {
-			query := dns.GetQuery()
-			ans := dns.GetIps()
-			if dns.GetQtypes()[0] == "Q" {
-				p.l.Sugar().Debugf("query from %s to %s: request %s\n", fl.GetIP().GetSource(), fl.GetIP().GetDestination(), query)
-			} else {
-				p.l.Sugar().Debugf("answer from %s to %s: result: %+v\n", fl.GetIP().GetSource(), fl.GetIP().GetDestination(), ans)
-			}
-		}
-	}
+	return nil
+}
 
-	if p.enricher != nil {
-		p.enricher.Write(ev)
-	} else {
-		p.l.Error("enricher is nil when writing event")
+func (p *Plugin) GetFlow(ctx context.Context, eventCh chan<- *v1.Event) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	err := p.StartStream(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to setup pktmon stream: %w", err)
 	}
 
-	// Write the event to the external channel.
-	if p.externalChannel != nil {
+	for {
```

Comment:
It seems like GetFlow functionality has changed? This infinite loop seems new, and the function now keeps getting flows until one of them is nil or an error is reached?

Reply:
yep exactly, all of the rpc stream management is here, and parent pktmon plugin is no longer responsible for grpc stream management